### PR TITLE
have cleanse plugin delete empty directories from the managed repos

### PR DESCRIPTION
**Pull request by mtbc**
After an image is deleted, its files are deleted from the managed repository but its empty directories are not deleted. Try importing some images, especially plates with subdirectory structure, in the client delete a subset of those images, then see if running cleanse as admin now deletes the corresponding empty directories.

--no-rebase

### CODE REVIEW

**Comment by ximenesuk** [comment]
So, I have deleted three images belonging to `user-2` and then I have run `cleanse` as three different users.
`user-4`
```
jrs-macbookpro:ome cblackburn$ omero -C admin cleanse ~/var/omero51
Server: [localhost:4064]
Username: [root]user-4
Password:
Created session a837acfd-fce3-4688-b8ea-7e4fac0c2eca (user-4@localhost:4064). Idle timeout: 10.0 min. Current group: read-only-1
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-03.708/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-08.226/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-11.294/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-13.220/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-16.097/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-57.944/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-17.808/
Traceback (most recent call last):
  File "/Users/cblackburn/Work/repos/ome/dist/bin/omero", line 125, in <module>
    rv = omero.cli.argv()
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1412, in argv
    cli.invoke(args[1:])
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 927, in invoke
    stop = self.onecmd(line, previous_args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1004, in onecmd
    self.execute(line, previous_args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1086, in execute
    args.func(args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/plugins/admin.py", line 1588, in cleanse
    dry_run=args.dry_run)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/util/cleanse.py", line 242, in cleanse
    delete_empty_dirs(proxy, root, client, dry_run)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/util/cleanse.py", line 268, in delete_empty_dirs
    raise Exception("failed delete: " + ce.err.message)
Exception: failed delete: not permitted to delete ome.model.core.OriginalFile[262]
```
nothing appears to have been deleted which makes sense. Though the output doesn't make that clear.
`user-2`
```
jrs-macbookpro:ome cblackburn$ omero -C admin cleanse ~/var/omero51
Server: [localhost:4064]
Username: [user-4]user-2
Password:
Created session cee2bc3e-f6fc-423d-9869-823e519e4546 (user-2@localhost:4064). Idle timeout: 10.0 min. Current group: private-1
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-13.220/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-57.944/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-17.808/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-03.708/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-08.226/
Traceback (most recent call last):
  File "/Users/cblackburn/Work/repos/ome/dist/bin/omero", line 125, in <module>
    rv = omero.cli.argv()
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1412, in argv
    cli.invoke(args[1:])
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 927, in invoke
    stop = self.onecmd(line, previous_args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1004, in onecmd
    self.execute(line, previous_args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/cli.py", line 1086, in execute
    args.func(args)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/plugins/admin.py", line 1588, in cleanse
    dry_run=args.dry_run)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/util/cleanse.py", line 242, in cleanse
    delete_empty_dirs(proxy, root, client, dry_run)
  File "/Users/cblackburn/Work/repos/ome/dist/lib/python/omero/util/cleanse.py", line 268, in delete_empty_dirs
    raise Exception("failed delete: " + ce.err.message)
Exception: failed delete: not permitted to delete ome.model.core.OriginalFile[262]
```
Again nothing is deleted which makes less sense if a user is allowed to cleanse their own space. The initial batch is a little different and again the output is not that helpful (to a user).

Finally, as `root`:
```
jrs-macbookpro:ome cblackburn$ omero -C admin cleanse ~/var/omero51
Server: [localhost:4064]
Username: [user-2]root
Password:
Created session 380dbf03-5f7c-4e57-9986-6ec2949a492e (root@localhost:4064). Idle timeout: 10.0 min. Current group: system
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-13.220/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-57.944/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-49-17.808/
```
Which does work and deletes the three expected  directories.

Maybe I need to understand what is happening a little more!

**Comment by mtbc** [comment]
Thank you for the report!

Yeah, it's really not designed for non-admins, it will just happen to work for them sometimes. So far it seems to be working about as well as I expected!

I suppose I could get the files' permissions objects to check `canDelete` from the Python client? If so, there might be an easy improvement I could make there for non-admins.

**Comment by ximenesuk** [comment]
Could it be just restricted to admins? I.e. not run at all as per some other plugins?

**Comment by ximenesuk** [comment]
I'm not sure how much difference this makes in a group environment. If on my local system I call `cleanse` as an ordinary user who has no data or has data but nothing to delete I see a list of `Removing` messages for non-empty directories of other members of my groups. Although this is followed by an exception it's just a single 'failed delete` message. If I was an ordinary user I would be panicking at this point that I had removed all but the last of the listed paths!

Interestingly if I run `cleanse` as an admin user with nothing to delete I don't see the paths of fellow group members.

So here, `user-2` has data but they see `user-6`'s paths:
```
jrs-macbookpro:ome cblackburn$ omero admin cleanse -C ~/var/omero51
Server: [localhost:4064]
Username: [user-2]
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-03.708/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-51-08.226/
Traceback (most recent call last):
...
Exception: failed delete: not permitted to delete ome.model.core.OriginalFile[262]
```
Here  `user-1` has no data but they see `user-2` and `user-6`'s paths:
```
jrs-macbookpro:ome cblackburn$ omero admin cleanse -C ~/var/omero51
Server: [localhost:4064]
Username: [user-3]user-1
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-6_8/2015-02/18/14-53-09.378/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-13.699/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-18.119/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-19.268/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-20.651/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-27.254/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-28.815/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-29.887/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-06.231/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-47-26.690/
Removing /Users/cblackburn/var/omero51/ManagedRepository/user-2_4/2015-02/18/14-48-10.541/
Traceback (most recent call last):
 ...
Exception: failed delete: not permitted to delete ome.model.core.OriginalFile[268]
```
But here `user-6` has data but they see nothing at all:
```
jrs-macbookpro:ome cblackburn$ omero admin cleanse -C ~/var/omero51
Server: [localhost:4064]
Username: [user-2]user-6
/Users/cblackburn/var/omero51/Pixels does not exist. Skipping...
```
This last result is the only one that makes sense to me, the user has nothing to delete so sees no messages.

Is there any way to either suppress the list for paths that don't belong to the user? Or provide something more than a single exception re-assuring the user that nothing has been deleted?

**Comment by ximenesuk** [comment]
Sorry, my fault! I had managed to not include the final commit in my libs!

**Comment by ximenesuk** [comment]
This now seems to work as expected. Users can cleanse their own empty directories. Admins can cleanse their own and other users' empty directories. There are no messages listing other directories as was seen before. And db relevant db rows are removed. Good to merge.

**Comment by ximenesuk** [comment]
Actually, is it worth updating the help here? It currently starts:
```
jrs-macbookpro:ome cblackburn$ omero admin cleanse --help
...
Remove binary data files from OMERO

Deleting an object from OMERO currently does not remove the binary data. Use
this command either manually or in a cron job periodically to remove Pixels
and other data.
```
This paragraph could probably be updated beyond has chenged just from this PR.

**Comment by mtbc** [comment]
Thank you again, @ximenesuk.

Hold off on merging: I thought of another improvement I want to try slipping in to this PR.

**Comment by mtbc** [comment]
The latest commit should remove the weirdness that directories would be deleted only if some other file were deletable by the user.

File: components/tools/OmeroPy/src/omero/util/cleanse.py [hunk 1]
```diff
@@ -226,6 +233,66 @@ def cleanse(data_dir, query_service, dry_run=False, config_service=None):
...
+            try:
+                client.waitOnCmd(handle, closehandle=True)
+            except omero.CmdError, ce:
+                if isinstance(ce.err, omero.cmd.GraphException):
+                    raise Exception("failed delete: " + ce.err.message)
+                else:
+                    raise Exception("failed: " + ce.err.name)
+
+
+def is_empty_dir(repo, directory, may_delete_dir, to_delete, dry_run):
```
Comment (ximenesuk) [inline components/tools/OmeroPy/src/omero/util/cleanse.py]: `dry_run` isn't needed as a parameter in this method call as far as I can see.

**Comment by ximenesuk** [comment]
Looks good functionally.

File: components/tools/OmeroPy/src/omero/util/cleanse.py (hunk 1 above)
Comment (mtbc) [inline components/tools/OmeroPy/src/omero/util/cleanse.py]: I wonder what I was using it for before.

Comment (ximenesuk) [inline components/tools/OmeroPy/src/omero/util/cleanse.py]: Maybe you originally deleted/listed while recursing?

**Comment by ximenesuk** [comment]
Good to merge.

**Pull request merged**